### PR TITLE
[ci] add versioning to CI workflow

### LINKED ISSUE #426: Add versioning to CI

### PR DESCRIPTION
**Pull request by simonvadee**
Fix #426

### CODE REVIEW

File: .github/workflows/server.yml [hunk 1]
```diff
@@ -31,8 +31,6 @@ jobs:
           yarn install
 
       - name: Build
-        env:
-          JWT_PRIVATE_KEY: none
         run: |
           cd server
```
Comment (Jasopaum) [inline .github/workflows/server.yml:35]: Isn't it possible to set the working directory somewhere for this job?

File: server/.env.docker [hunk 2]
```diff
@@ -2,4 +2,5 @@ PORT=1000
 POSTGRES_URL=postgresql://prisma:prisma@postgres:5432/prisma?schema=pyrog
 APP_SECRET=deadbeefdeadbeefdeadbeefdeadbeef
 FHIR_API_URL=http://reverse-proxy/api
-REDIS_URL=redis://redis:6379
\ No newline at end of file
+REDIS_URL=redis://redis:6379
+ENV=test
```
Comment (Jasopaum) [inline server/.env.docker:6]: Do we want to keep that?

File: server/docker-entrypoint.sh [hunk 3]
```diff
@@ -0,0 +1,17 @@
+#!/bin/sh
+
+# Exit immediately if a command exits with a non-zero status.
+set -e
+
+echo "test"
```
Comment (Jasopaum) [inline server/docker-entrypoint.sh:6]: Don't forget this one

Comment (simonvadee) [reply server/docker-entrypoint.sh:6]: thx

File: server/.env.docker (hunk 2 above)
Comment (simonvadee) [reply server/.env.docker:6]: yes, I don't think we want to setup the whole hydra mess when running locally in docker-compose (see #429)

File: .github/workflows/server.yml (hunk 1 above)
Comment (simonvadee) [reply .github/workflows/server.yml:38]: yes! here you go

File: .github/workflows/client.yml [hunk 4]
```diff
@@ -1,86 +1,83 @@
...
 
-      - name: Tag docker image with hash
-        run: |
-          docker tag $IMAGE:${{ matrix.environment.name }} $IMAGE:${{ matrix.environment.name }}-$GITHUB_SHA
+          # Has this workflow been triggered by an app tag ?
+          if [[ "${GITHUB_REF}" == refs/tags/app/* ]]; then
+            VERSION=${GITHUB_REF#refs/tags/app/}
+            # Is the tag on master and does it follow numeric versioning ?
+            if [[ "${{ github.event.base_ref }}" == refs/heads/master ]] && [[ "${VERSION}" =~ ^v[0-9]{1,3}(\.[0-9]{1,3}(\.[0-9]{1,3})?)?$ ]]; then
+              TAGS="${TAGS},${IMAGE}:latest"
```
Comment (vmttn) [inline .github/workflows/client.yml:45]: I suggest we strip the leading `v` from this tag. It's standard for numerically versioned docker image.
Cf arkhn/warehouse-api/pull/160

```suggestion
              TAGS="${TAGS},${IMAGE}:latest"
              VERSION="${VERSION#v}"
```

File: client/docker-compose.yml [hunk 5]
```diff
@@ -1,7 +1,7 @@
 version: "3.7"
 services:
-  pyrog_client:
-    image: arkhn/pyrog-client:latest
+  client:
+    image: pyrog-client:latest
```
Comment (vmttn) [inline client/docker-compose.yml:4]: Remove `latest` ?

File: server/.dockerignore [hunk 6]
```diff
@@ -11,4 +11,5 @@
 !tsconfig.build.json
 !prisma
 !scripts
-!.env*
\ No newline at end of file
+!.env*
+!docker-entrypoint.sh
```
Comment (vmttn) [inline server/.dockerignore:15]: Can we invert this ? Include everything but the commonly excluded files (node_modules, transpiled files, etc) Less likely to change.

File: server/Dockerfile [hunk 7]
```diff
@@ -21,9 +20,18 @@ COPY src src
...
+COPY ["package.json",  "tsconfig.json", "tsconfig.build.json", "docker-entrypoint.sh", "/srv/"]
+COPY prisma prisma
 COPY scripts scripts
 
-# Start the app
-ENV NODE_PATH dist/
+COPY --from=build-image /build/node_modules ./node_modules
+COPY --from=build-image /build/dist ./dist
+
+EXPOSE 1000
```
Comment (vmttn) [inline server/Dockerfile:34]: :+1:

File: server/docker-compose.yml [hunk 8]
```diff
@@ -2,20 +2,15 @@ version: '3.7'
 
 services:
   pyrog:
-    image: arkhn/pyrog-server:latest
+    image: pyrog-server:latest
```
Comment (vmttn) [inline server/docker-compose.yml:5]: Same, remove `latest` ?

File: .github/workflows/server.yml [hunk 9]
```diff
@@ -25,58 +28,69 @@ jobs:
...
+          if [[ "${GITHUB_REF}" == refs/tags/api/* ]]; then
+            VERSION=${GITHUB_REF#refs/tags/api/}
 
-      - name: Build
-        run: |
-          cd server/
-          yarn build:docker
+            # Is the tag on master and does it follow numeric versioning ?
+            if [[ "${{ github.event.base_ref }}" == refs/heads/master ]] && [[ "${VERSION}" =~ ^v[0-9]{1,3}(\.[0-9]{1,3}(\.[0-9]{1,3})?)?$ ]]; then
+              TAGS="${TAGS},${IMAGE}:latest"
```
Comment (vmttn) [inline .github/workflows/server.yml:56]:
```suggestion
              TAGS="${TAGS},${IMAGE}:latest"
              VERSION="${VERSION#v}"
```

File: client/.dockerignore [hunk 10]
```diff
@@ -10,3 +10,4 @@
 !src
 !public
 !.env*
+ !docker-entrypoint.sh
```
Comment (vmttn) [inline client/.dockerignore:13]: Same. Can we invert this ? Include everything but the commonly excluded files (node_modules, transpiled files, etc) Less likely to change.

**Review by vmttn — CHANGES_REQUESTED** [review]

File: client/docker-compose.yml (hunk 5 above)
Comment (simonvadee) [reply client/docker-compose.yml:4]: done

File: server/docker-compose.yml (hunk 8 above)
Comment (simonvadee) [reply server/docker-compose.yml:5]: done

File: server/.dockerignore (hunk 6 above)
Comment (simonvadee) [reply server/.dockerignore:15]: I prefer using the .dockerignore as a whitelist to be explicit about what's in the image. Yes it's more likely to change, but it's easy to spot that a file is missing in the docker context when trying to build the image.

File: client/.dockerignore (hunk 10 above)
Comment (simonvadee) [reply client/.dockerignore:13]: same

**Review by vmttn — APPROVED** [review]

**Pull request merged**